src/main/java/com/fcgl/madrid/points/service/UserActionService.java

```diff
@@ -1,22 +1,70 @@
 package com.fcgl.madrid.points.service;
 
+import com.fcgl.madrid.points.model.Action;
+import com.fcgl.madrid.points.model.Trophy;
 import com.fcgl.madrid.points.model.UserAction;
+import com.fcgl.madrid.points.model.UserTrophy;
+import com.fcgl.madrid.points.repository.ActionRepository;
+import com.fcgl.madrid.points.repository.TrophyRepository;
 import com.fcgl.madrid.points.repository.UserActionRepository;
+import com.fcgl.madrid.points.repository.UserTrophyRepository;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.ResponseEntity;
 import org.springframework.stereotype.Service;
 
 import java.util.List;
 
 @Service
 public class UserActionService {
     private UserActionRepository userActionRepository;
+    private ActionRepository actionRepository; // Is this the correct way to access the action table?
+    private TrophyRepository trophyRepository;
+    private UserTrophyRepository userTrophyRepository;
 
     @Autowired
-    public UserActionService(UserActionRepository userActionRepository) {
+    public UserActionService(UserActionRepository userActionRepository,
+                             ActionRepository actionRepository,
+                             TrophyRepository trophyRepository,
+                             UserTrophyRepository userTrophyRepository) {
         this.userActionRepository = userActionRepository;
+        this.actionRepository = actionRepository;
+        this.trophyRepository = trophyRepository;
+        this.userTrophyRepository = userTrophyRepository;
     }
 
     public List<UserAction> findAll() {
         return userActionRepository.findAll();
     }
+
+    public ResponseEntity<Trophy> addUserAction(Long userID, Long actionID) {
+
+        // Check that the action exists
+        Action action = actionRepository.findById(actionID).orElse(null);
+        if (action == null) {
+            return ResponseEntity.badRequest().body(null);
+        }
+
+        UserAction userAction = userActionRepository.findByUserIdAndAction(userID, action).orElse(null);
+        if (userAction == null) {
+            // We haven't seen this action for this user,
+            // Let's create a new entry with count equal to 1
+            userAction = new UserAction(userID, action, 1);
+            userActionRepository.save(userAction);
```

Comment:
Do we need to catch exceptions here? JPA default exception handler?

Reply:
Yeah, I was looking into that for the other teams because I wanted to update the way we did the service logic for `resilience4j`

Anytime we do anything that interacts with the database we should handle `JDBCConnectionException`. When there is a `JDBCConnectionException` and perhaps also `QueryTimeoutException` exception I think it would be good for that to be done through `resilience4j` so that we can stop requests to the database while the database is having issues

Here are some exceptions that could be thrown: https://www.baeldung.com/hibernate-exceptions

The only exception that we have been handling has been the `ConstraintViolationException`. An implementation of that can be found here: https://github.com/fcgl/point-system/pull/7/files#diff-8764bc014ac88907822362160acdf126R53